controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,891 @@
+			virtualNetworkResourceGroup,
+			spec.VirtualNetwork,
+			spec.Subnet,
+		))
+
+		networkProfile.DNSServiceIP = to.StringPtr(spec.NetworkDNSServiceIP)
+		networkProfile.DockerBridgeCidr = to.StringPtr(spec.NetworkDockerBridgeCIDR)
+		networkProfile.ServiceCidr = to.StringPtr(spec.NetworkServiceCIDR)
+
+		if spec.NetworkPlugin == "" {
+			networkProfile.NetworkPlugin = containersv211.Kubenet
+		} else {
+			networkProfile.NetworkPlugin = containersv211.NetworkPlugin(spec.NetworkPlugin)
+		}
+
+		// if network plugin is 'kubenet', set PodCIDR
+		if networkProfile.NetworkPlugin == containersv211.Azure {
+			networkProfile.PodCidr = to.StringPtr(spec.NetworkPodCIDR)
+		}
+
+		if spec.LoadBalancerSku != "" {
+			loadBalancerSku := containersv211.LoadBalancerSku(spec.LoadBalancerSku)
+			networkProfile.LoadBalancerSku = loadBalancerSku
+		}
+
+		if spec.NetworkPolicy != "" {
+			networkProfile.NetworkPolicy = containersv211.NetworkPolicy(spec.NetworkPolicy)
+		}
+	}
+
+	var agentPoolProfiles []containersv211.ManagedClusterAgentPoolProfile
+	if hasAgentPoolProfile(spec) {
+		for _, np := range spec.NodePools {
+			if np.OrchestratorVersion == "" {
+				np.OrchestratorVersion = spec.KubernetesVersion
+			}
+			agentProfile := containersv211.ManagedClusterAgentPoolProfile{
+				Name:                to.StringPtr(np.Name),
+				Count:               to.Int32Ptr(np.Count),
+				MaxPods:             to.Int32Ptr(np.MaxPods),
+				OsDiskSizeGB:        to.Int32Ptr(np.OsDiskSizeGB),
+				OsType:              containersv211.OSType(np.OsType),
+				VMSize:              containersv211.VMSizeTypes(np.VMSize),
+				Mode:                containersv211.AgentPoolMode(np.Mode),
+				OrchestratorVersion: to.StringPtr(np.OrchestratorVersion),
+			}
+			if len(np.AvailabilityZones) > 0 {
+				agentProfile.AvailabilityZones = to.StringSlicePtr(np.AvailabilityZones)
+			}
+			if hasCustomVirtualNetwork(spec) {
+				agentProfile.VnetSubnetID = vmNetSubnetID
+			}
+			agentPoolProfiles = append(agentPoolProfiles, agentProfile)
+		}
+	} else {
+		agentProfile := containersv211.ManagedClusterAgentPoolProfile{
+			Name:         to.StringPtr("system"),
+			Count:        to.Int32Ptr(1),
+			MaxPods:      to.Int32Ptr(110),
+			OsDiskSizeGB: to.Int32Ptr(128),
+			OsType:       containersv211.Linux,
+			VMSize:       containersv211.StandardDS2V2,
+			Mode:         containersv211.System,
+		}
+		agentPoolProfiles = append(agentPoolProfiles, agentProfile)
+	}
+
+	var linuxProfile *containersv211.LinuxProfile
+	if hasLinuxProfile(spec) {
+		linuxProfile = &containersv211.LinuxProfile{
+			AdminUsername: to.StringPtr(spec.LinuxAdminUsername),
+			SSH: &containersv211.SSHConfiguration{
+				PublicKeys: &[]containersv211.SSHPublicKey{
+					{
+						KeyData: to.StringPtr(spec.LinuxSSHPublicKeyContents),
+					},
+				},
+			},
+		}
+	}
+
+	clientId, clientSecret, err := getSecrets(secretsCache, spec.AzureCredentialSecret)
+	if err != nil {
+		return c, fmt.Errorf("couldn't get secrets")
```

Comment:
nit: consider including the error from `getSecrets` to give more context.

Reply:
I included error message, thanks